File: clients/go/model/debug_data.go
package model

import "fmt"
import "io"
import . "komus24/stream"

// TODO - Document
type DebugData interface {
    // Write DebugData to writer
    Write(writer io.Writer)

    // Get string representation of DebugData
    String() string
}

// Read DebugData from reader
func ReadDebugData(reader io.Reader) DebugData {
    switch ReadInt32(reader) {
    case 0:
        return ReadDebugDataCircle(reader)
    }
    panic("Unexpected tag value")
}

// TODO - Document
type DebugDataCircle struct {
    // TODO - Document
    Pos Vec2Float64
    // TODO - Document
    Radius float64
}

func NewDebugDataCircle(pos Vec2Float64, radius float64) DebugDataCircle {
    return DebugDataCircle {
        Pos: pos,
        Radius: radius,
    }
}

// Read Circle from reader
func ReadDebugDataCircle(reader io.Reader) DebugDataCircle {
    var pos Vec2Float64
    pos = ReadVec2Float64(reader)
    var radius float64
    radius = ReadFloat64(reader)
    return DebugDataCircle {
        Pos: pos,
        Radius: radius,
    }
}

// Write Circle to writer
func (debugDataCircle DebugDataCircle) Write(writer io.Writer) {
    WriteInt32(writer, 0)
    pos := debugDataCircle.Pos
    pos.Write(writer)
    radius := debugDataCircle.Radius
    WriteFloat64(writer, radius)
}

// Get string representation of Circle
func (debugDataCircle DebugDataCircle) String() string {
    stringResult := "{ "
    stringResult += "Pos: "
    pos := debugDataCircle.Pos
    stringResult += pos.String()
    stringResult += ", "
    stringResult += "Radius: "
    radius := debugDataCircle.Radius
    stringResult += fmt.Sprint(radius)
    stringResult += " }"
    return stringResult
}

File: clients/python_pypy/model/debug_data.py
from model.vec2_double import Vec2Double
from stream_wrapper import StreamWrapper

class DebugData:
    """TODO - Document"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DebugData":
        """Read DebugData from input stream
        """
        tag = stream.read_int()
        if tag == Circle.TAG:
            return DebugData.Circle.read_from(stream)
        raise Exception("Unexpected tag value")

class Circle(DebugData):
    """TODO - Document"""

    TAG = 0

    __slots__ = ("pos","radius",)

    pos: Vec2Double
    radius: float

    def __init__(self, pos: Vec2Double, radius: float):
        self.pos = pos
        """TODO - Document"""
        self.radius = radius
        """TODO - Document"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Circle":
        """Read Circle from input stream
        """
        pos = Vec2Double.read_from(stream)
        radius = stream.read_double()
        return Circle(pos, radius)
    
    def write_to(self, stream: StreamWrapper):
        """Write Circle to output stream
        """
        stream.write_int(self.TAG)
        self.pos.write_to(stream)
        stream.write_double(self.radius)
    
    def __repr__(self):
        return "Circle(" + \
            repr(self.pos) + \
            ", " + \
            repr(self.radius) + \
            ")"

DebugData.Circle = Circle